src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -921,53 +1203,27 @@ HeapWord* ShenandoahFreeSet::allocate_contiguous(ShenandoahAllocRequest& req) {
     r->set_affiliation(req.affiliation());
     r->set_update_watermark(r->bottom());
     r->set_top(r->bottom() + used_words);
-
-    // While individual regions report their true use, all humongous regions are marked used in the free set.
-    _free_sets.remove_from_free_sets(r->index());
   }
   generation->increase_affiliated_region_count(num);
+  if (remainder != 0) {
+    // Record this remainder as allocation waste
+    _heap->notify_mutator_alloc_words(ShenandoahHeapRegion::region_size_words() - remainder, true);
+  }
+
+  // retire_range_from_partition() will adjust bounds on Mutator free set if appropriate
+  _partitions.retire_range_from_partition(ShenandoahFreeSetPartitionId::Mutator, beg, end);
 
   size_t total_humongous_size = ShenandoahHeapRegion::region_size_bytes() * num;
-  _free_sets.increase_used(Mutator, total_humongous_size);
-  _free_sets.assert_bounds();
+  _partitions.increase_used(ShenandoahFreeSetPartitionId::Mutator, total_humongous_size);
+  _partitions.assert_bounds();
   req.set_actual_size(words_size);
   if (remainder != 0) {
     req.set_waste(ShenandoahHeapRegion::region_size_words() - remainder);
   }
   return _heap->get_region(beg)->bottom();
 }
 
-// Returns true iff this region is entirely available, either because it is empty() or because it has been found to represent
-// immediate trash and we'll be able to immediately recycle it.  Note that we cannot recycle immediate trash if
-// concurrent weak root processing is in progress.
-bool ShenandoahFreeSet::can_allocate_from(ShenandoahHeapRegion *r) const {
-  return r->is_empty() || (r->is_trash() && !_heap->is_concurrent_weak_root_in_progress());
-}
-
-bool ShenandoahFreeSet::can_allocate_from(size_t idx) const {
-  ShenandoahHeapRegion* r = _heap->get_region(idx);
-  return can_allocate_from(r);
-}
-
-size_t ShenandoahFreeSet::alloc_capacity(size_t idx) const {
-  ShenandoahHeapRegion* r = _heap->get_region(idx);
-  return alloc_capacity(r);
-}
-
-size_t ShenandoahFreeSet::alloc_capacity(ShenandoahHeapRegion *r) const {
-  if (r->is_trash()) {
-    // This would be recycled on allocation path
-    return ShenandoahHeapRegion::region_size_bytes();
-  } else {
-    return r->free();
-  }
-}
-
-bool ShenandoahFreeSet::has_alloc_capacity(ShenandoahHeapRegion *r) const {
-  return alloc_capacity(r) > 0;
-}
-
-void ShenandoahFreeSet::try_recycle_trashed(ShenandoahHeapRegion *r) {
+void ShenandoahFreeSet::try_recycle_trashed(ShenandoahHeapRegion *r, bool is_generational) {
```

Comment:
`is_generational` looks unused.

Reply:
Thanks.  Removing this unneeded argument.